File: app/lib/services/dufs_client.dart
import 'dart:convert';
import 'dart:io' show HttpDate;

import 'package:dufs_client/models/dir_entry.dart';
import 'package:dufs_client/models/media_meta.dart';
import 'package:dufs_client/util/paths.dart' as paths;
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart';

/// Client for the dufs server over HTTP + WebDAV, adding HTTP Basic auth to
/// every request (the mobile app supplies credentials explicitly, unlike the
/// same-origin web UI where the browser handles auth).
class DufsClient {
  /// Creates a client for [baseUrl] with the given credentials. A custom
  /// [httpClient] can be injected for testing.
  DufsClient({
    required this.baseUrl,
    required this.username,
    required this.password,
    http.Client? httpClient,
  }) : _http = httpClient ?? http.Client();

  /// Base URL, e.g. `https://host` (a trailing slash is tolerated).
  final String baseUrl;

  /// dufs web username.
  final String username;

  /// dufs web password.
  final String password;

  final http.Client _http;

  /// HTTP Basic auth headers — also passed to `Image.network` and the video
  /// player so media requests are authenticated too.
  Map<String, String> get authHeaders => <String, String>{
        'authorization':
            'Basic ${base64Encode(utf8.encode('$username:$password'))}',
      };

  /// Builds the authenticated absolute URL of a cloud [path].
  Uri fileUri(String path) => _uri(path);

  /// URL of the generated thumbnail for a media entry (see
  /// `generate_thumbnails.sh`); falls back to an icon when it 404s.
  Uri thumbUri(String path) => _uri('/.thumbs${paths.normalize(path)}.jpg');

  Uri _uri(String path) {
    final root = baseUrl.endsWith('/')
        ? baseUrl.substring(0, baseUrl.length - 1)
        : baseUrl;
    final encoded = path
        .split('/')
        .where((s) => s.isNotEmpty)
        .map(Uri.encodeComponent)
        .join('/');
    return Uri.parse('$root/$encoded');
  }

  /// Lists directory [dirPath] via WebDAV PROPFIND.
  Future<List<DirEntry>> list(String dirPath) async {
    final request = http.Request('PROPFIND', _uri(dirPath))
      ..headers.addAll(<String, String>{...authHeaders, 'depth': '1'});
    final response = await _http.send(request);
    final body = await response.stream.bytesToString();
    if (response.statusCode >= 400) {
      throw Exception('PROPFIND $dirPath -> ${response.statusCode}');
    }
    return parsePropfind(body, dirPath);
  }

  /// Uploads [bytes] to [dirPath]/[name] with a WebDAV PUT.
  Future<void> upload(String dirPath, String name, List<int> bytes) async {
    final target = dirPath.endsWith('/') ? '$dirPath$name' : '$dirPath/$name';
    final response = await _http.put(_uri(target),
        headers: authHeaders, body: bytes);
    if (response.statusCode >= 400) {
      throw Exception('PUT $target -> ${response.statusCode}');
    }
  }

  /// Downloads the raw bytes of a file.
  Future<List<int>> download(String path) async {
    final response = await _http.get(_uri(path), headers: authHeaders);
    if (response.statusCode >= 400) {
      throw Exception('GET $path -> ${response.statusCode}');
    }
    return response.bodyBytes;
  }

  /// Deletes a file or directory.
  Future<void> delete(String path) async {
    final response = await _http.delete(_uri(path), headers: authHeaders);
    if (response.statusCode >= 400) {
      throw Exception('DELETE $path -> ${response.statusCode}');
    }
  }

  /// Creates a directory at [path] (WebDAV MKCOL).
  Future<void> createDir(String path) async {
    final request = http.Request('MKCOL', _uri(path))
      ..headers.addAll(authHeaders);
    final response = await _http.send(request);
    if (response.statusCode >= 400) {
      throw Exception('MKCOL $path -> ${response.statusCode}');
    }
  }

  /// Moves [fromPath] into directory [destDir], keeping its base name (MOVE).
  Future<void> move(String fromPath, String destDir) async {
    final dest = paths.joinPath(destDir, paths.basename(fromPath));
    final request = http.Request('MOVE', _uri(fromPath))
      ..headers.addAll(<String, String>{
        ...authHeaders,
        'destination': _uri(dest).toString(),
        'overwrite': 'F',
      });
    final response = await _http.send(request);
    if (response.statusCode >= 400) {
      throw Exception('MOVE $fromPath -> ${response.statusCode}');
    }
  }

  /// Renames [path] to [newName], keeping it in the same directory (MOVE).
  Future<void> rename(String path, String newName) async {
    final dest = paths.joinPath(paths.parentPath(path), newName);
    final request = http.Request('MOVE', _uri(path))
      ..headers.addAll(<String, String>{
        ...authHeaders,
        'destination': _uri(dest).toString(),
        'overwrite': 'F',
      });
    final response = await _http.send(request);
    if (response.statusCode >= 400) {
      throw Exception('MOVE $path -> ${response.statusCode}');
    }
  }

  /// Reads a text file's contents.
  Future<String> readText(String path) async {
    final response = await _http.get(_uri(path), headers: authHeaders);
    if (response.statusCode >= 400) {
      throw Exception('GET $path -> ${response.statusCode}');
    }
    return response.body;
  }

  /// Writes [content] to a text file (WebDAV PUT).
  Future<void> writeText(String path, String content) async {
    final response =
        await _http.put(_uri(path), headers: authHeaders, body: content);
    if (response.statusCode >= 400) {
      throw Exception('PUT $path -> ${response.statusCode}');
    }
  }

  /// Fetches the server metadata index (`/.meta/index.json`); resolves to an
  /// empty index on a missing file, non-ok status, network failure, or bad
  /// JSON — the index is an optional enrichment, never a hard dependency.
  Future<MetaIndex> fetchMeta() async {
    try {
      final response =
          await _http.get(_uri('/.meta/index.json'), headers: authHeaders);
      if (response.statusCode >= 400) return <String, MediaMeta>{};
      return metaIndexFromJson(jsonDecode(response.body));
    } on Exception {
      return <String, MediaMeta>{};
    }
  }

  /// Releases the underlying HTTP client.
  void close() => _http.close();
}

/// Parses a dufs WebDAV multistatus body into the entries under [dirPath],
/// dropping the directory's own self entry. Exposed for testing.
List<DirEntry> parsePropfind(String xmlBody, String dirPath) {
  final self = _normalize(dirPath);
  final document = XmlDocument.parse(xmlBody);
  final entries = <DirEntry>[];
  for (final response in document.findAllElements('response', namespace: '*')) {
    final href = response
        .findElements('href', namespace: '*')
        .firstOrNull
        ?.innerText;
    if (href == null || href.isEmpty) continue;
    final path = _normalize(Uri.decodeComponent(href));
    if (path == self) continue;
    final isDir = response
        .findAllElements('collection', namespace: '*')
        .isNotEmpty;
    final sizeText = response
        .findAllElements('getcontentlength', namespace: '*')
        .firstOrNull
        ?.innerText;
    final mtimeText = response
        .findAllElements('getlastmodified', namespace: '*')
        .firstOrNull
        ?.innerText;
    entries.add(
      DirEntry(
        name: _basename(path),
        path: path,
        kind: isDir ? EntryKind.dir : EntryKind.file,
        size: int.tryParse(sizeText ?? '') ?? 0,
        mtimeMs: _parseHttpDateMs(mtimeText),
      ),
    );
  }
  entries.sort((a, b) {
    if (a.isDir != b.isDir) return a.isDir ? -1 : 1;
    return a.name.toLowerCase().compareTo(b.name.toLowerCase());
  });
  return entries;
}

int _parseHttpDateMs(String? text) {
  if (text == null || text.isEmpty) return 0;
  try {
    return HttpDate.parse(text).millisecondsSinceEpoch;
  } on Exception {
    return 0;
  }
}

String _normalize(String path) {
  final parts = path.split('/').where((p) => p.isNotEmpty && p != '.');
  return '/${parts.join('/')}';
}

String _basename(String path) {
  final n = _normalize(path);
  if (n == '/') return '/';
  return n.substring(n.lastIndexOf('/') + 1);
}

File: web/src/api/dufs-client.ts
import type {
  DirEntry,
  EntryKind,
  MetaIndex,
  SubtitleManifest,
} from "./types.ts";
import {
  basename,
  encodePath,
  joinPath,
  normalize,
  parentPath,
} from "../lib/paths.ts";

const DAV_NS = "DAV:";

/** Client over the dufs WebDAV + HTTP API (same-origin; browser supplies auth). */
export interface DufsClient {
  /** List a directory via WebDAV PROPFIND (works under dufs `render-spa`). */
  list(dirPath: string): Promise<DirEntry[]>;
  /** Same-origin URL to fetch/stream a file's raw bytes. */
  fileUrl(path: string): string;
  /** URL of the generated thumbnail for a media entry (see generate_thumbnails.sh). */
  thumbUrl(path: string): string;
  /** URL that downloads a directory as a zip (dufs `?zip`). */
  zipUrl(dirPath: string): string;
  upload(dirPath: string, file: File): Promise<void>;
  remove(path: string): Promise<void>;
  /** Create a directory at `path` (WebDAV MKCOL). */
  createDir(path: string): Promise<void>;
  /** Move `fromPath` into directory `destDir`, keeping its base name (MOVE). */
  move(fromPath: string, destDir: string): Promise<void>;
  /** Rename `path` to `newName`, keeping it in the same directory (MOVE). */
  rename(path: string, newName: string): Promise<void>;
  readText(path: string): Promise<string>;
  writeText(path: string, content: string): Promise<void>;
  /** Fetch the server metadata index; resolves to `{}` when absent. */
  fetchMeta(): Promise<MetaIndex>;
  /** The subtitle manifest in `dir` (a video's `subtitlesPath`), or null when
   * it is absent or unreadable. Tolerant by the same contract as fetchMeta. */
  fetchSubtitleManifest(dir: string): Promise<SubtitleManifest | null>;
  /** Download a file's raw bytes (used to build multi-file zips). */
  downloadBytes(path: string): Promise<Uint8Array>;
}

/** Parse a dufs PROPFIND multistatus XML body into entries under `dirPath`. */
export function parsePropfind(xml: string, dirPath: string): DirEntry[] {
  const doc = new DOMParser().parseFromString(xml, "application/xml");
  const self = normalize(dirPath);
  const responses = Array.from(doc.getElementsByTagNameNS(DAV_NS, "response"));
  const entries: DirEntry[] = [];
  for (const res of responses) {
    const hrefEl = res.getElementsByTagNameNS(DAV_NS, "href")[0];
    const rawHref = hrefEl?.textContent;
    // TypeScript 6 types `textContent` as `string`, so the optional chain
    // above yields `string | undefined` and an explicit `=== null` arm is
    // now unreachable. A falsy check still covers empty as it always did.
    if (rawHref === undefined || rawHref === "") continue;
    // href is URL-encoded and absolute; decode and normalize.
    const path = normalize(decodeURIComponent(rawHref));
    if (path === self) continue; // skip the directory itself
    const isDir =
      res.getElementsByTagNameNS(DAV_NS, "collection").length > 0;
    const kind: EntryKind = isDir ? "dir" : "file";
    const sizeText = res
      .getElementsByTagNameNS(DAV_NS, "getcontentlength")[0]
      ?.textContent;
    const size = sizeText !== undefined ? Number(sizeText) : 0;
    const mtimeText = res
      .getElementsByTagNameNS(DAV_NS, "getlastmodified")[0]
      ?.textContent;
    const mtimeMs = mtimeText !== undefined ? Date.parse(mtimeText) || 0 : 0;
    entries.push({
      name: basename(path),
      path,
      kind,
      size: Number.isFinite(size) ? size : 0,
      mtimeMs,
    });
  }
  return entries;
}

/** Sort: directories first, then by name (locale, case-insensitive). */
export function sortEntries(entries: readonly DirEntry[]): DirEntry[] {
  return [...entries].sort((a, b) => {
    if (a.kind !== b.kind) return a.kind === "dir" ? -1 : 1;
    return a.name.localeCompare(b.name, undefined, { sensitivity: "base" });
  });
}

export function createDufsClient(fetchImpl: typeof fetch = fetch): DufsClient {
  // `method` is always supplied by callers, so the error message and the fetch
  // options never need to fall back — keeping it a required arg avoids an
  // unreachable default branch.
  async function request(
    method: string,
    path: string,
    init: RequestInit = {},
  ): Promise<Response> {
    const res = await fetchImpl(encodePath(path), {
      method,
      credentials: "same-origin",
      ...init,
    });
    if (!res.ok) {
      throw new Error(`${method} ${path} → ${res.status}`);
    }
    return res;
  }

  return {
    async list(dirPath) {
      const res = await request("PROPFIND", dirPath, {
        headers: { Depth: "1" },
      });
      return sortEntries(parsePropfind(await res.text(), dirPath));
    },
    fileUrl(path) {
      return encodePath(path);
    },
    thumbUrl(path) {
      return encodePath(`/.thumbs${normalize(path)}.jpg`);
    },
    zipUrl(dirPath) {
      return `${encodePath(dirPath)}?zip`;
    },
    async upload(dirPath, file) {
      await request("PUT", joinPath(dirPath, file.name), { body: file });
    },
    async remove(path) {
      await request("DELETE", path);
    },
    async createDir(path) {
      await request("MKCOL", path);
    },
    async move(fromPath, destDir) {
      const dest = joinPath(destDir, basename(fromPath));
      await request("MOVE", fromPath, {
        headers: { Destination: encodePath(dest), Overwrite: "F" },
      });
    },
    async rename(path, newName) {
      const dest = joinPath(parentPath(path), newName);
      await request("MOVE", path, {
        headers: { Destination: encodePath(dest), Overwrite: "F" },
      });
    },
    async readText(path) {
      return (await request("GET", path)).text();
    },
    async writeText(path, content) {
      await request("PUT", path, { body: content });
    },
    async fetchMeta() {
      // Tolerant by contract: a missing index, a non-ok status, a network
      // failure (offline), or malformed JSON all resolve to {} — the index is
      // an optional enrichment, never a hard dependency of the listing.
      try {
        const res = await fetchImpl(encodePath("/.meta/index.json"), {
          credentials: "same-origin",
        });
        if (!res.ok) return {};
        const data: unknown = await res.json();
        return isMetaIndex(data) ? data.entries : {};
      } catch {
        return {};
      }
    },
    async fetchSubtitleManifest(dir) {
      // Same tolerance as fetchMeta: subtitles are an enrichment, and a video
      // whose manifest is missing simply has no embedded tracks to offer.
      try {
        const res = await fetchImpl(encodePath(`${dir}/tracks.json`), {
          credentials: "same-origin",
        });
        if (!res.ok) return null;
        const data: unknown = await res.json();
        return isSubtitleManifest(data) ? data : null;
      } catch {
        return null;
      }
    },
    async downloadBytes(path) {
      const res = await request("GET", path);
      return new Uint8Array(await res.arrayBuffer());
    },
  };
}

interface MetaFile {
  readonly entries: MetaIndex;
}

/** Narrow parsed JSON to the metadata index shape (tolerant of extra fields). */
function isSubtitleManifest(data: unknown): data is SubtitleManifest {
  if (typeof data !== "object" || data === null) return false;
  return (
    "tracks" in data &&
    Array.isArray(data.tracks) &&
    "fonts" in data &&
    Array.isArray(data.fonts)
  );
}

function isMetaIndex(data: unknown): data is MetaFile {
  if (typeof data !== "object" || data === null || !("entries" in data)) {
    return false;
  }
  const { entries } = data;
  return typeof entries === "object" && entries !== null;
}
